server/image_pull.go

```diff
@@ -80,22 +86,84 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 			} else if tmpImgConfigDigest.String() == storedImage.ConfigDigest.String() {
 				log.Debugf(ctx, "image %s already in store, skipping pull", img)
 				pulled = img
+
+				// Skipped bytes metrics
+				if storedImage.Size != nil {
+					counter, err := metrics.CRIOImagePullsByNameSkipped.GetMetricWithLabelValues(img)
+					if err != nil {
+						log.Warnf(ctx, "Unable to write image pull name (skipped) metrics: %v", err)
+					} else {
+						counter.Add(float64(*storedImage.Size))
+					}
+				}
+
 				break
 			}
 			log.Debugf(ctx, "image in store has different ID, re-pulling %s", img)
 		}
 
-		_, err = s.StorageImageServer().PullImage(s.systemContext, img, &copy.Options{
-			SourceCtx:        &sourceCtx,
-			DestinationCtx:   s.systemContext,
-			OciDecryptConfig: dcc,
-		})
+		// Pull by collecting progress metrics
+		overallSize, err := imageSize(tmpImg)
 		if err != nil {
+			log.Infof(ctx, "%v, skipping this metric", err)
+		}
+
+		wg := &sync.WaitGroup{}
+		progress := make(chan types.ProgressProperties)
+		wg.Add(1)
+		go func() {
+			_, err := s.StorageImageServer().PullImage(s.systemContext, img, &copy.Options{
+				SourceCtx:        &sourceCtx,
+				DestinationCtx:   s.systemContext,
+				OciDecryptConfig: dcc,
+				ProgressInterval: time.Second,
+				Progress:         progress,
+			})
+			if err != nil {
+				pullError <- err
+			}
+			wg.Done()
+		}()
+		go func() {
+			for p := range progress {
+				log.Debugf(ctx, "ImagePull (%v): %s (%s): %v bytes (%.2f%%)",
+					p.Event, img, p.Artifact.Digest, p.Offset,
+					float64(p.Offset)/float64(p.Artifact.Size)*100,
+				)
+
+				// Metrics for every digest
+				digestCounter, err := metrics.CRIOImagePullsByDigest.GetMetricWithLabelValues(
+					img, p.Artifact.Digest.String(), p.Artifact.MediaType,
+					fmt.Sprintf("%d", p.Artifact.Size),
+				)
+				if err != nil {
+					log.Warnf(ctx, "Unable to write image pull digest metrics: %v", err)
+				} else {
+					digestCounter.Add(float64(p.OffsetUpdate))
+				}
+
+				// Metrics for the overall image
+				if overallSize > 0 {
+					nameCounter, err := metrics.CRIOImagePullsByName.GetMetricWithLabelValues(
+						img, fmt.Sprintf("%d", overallSize),
+					)
+					if err != nil {
+						log.Warnf(ctx, "Unable to write image pull name metrics: %v", err)
+					} else {
+						nameCounter.Add(float64(p.OffsetUpdate))
+					}
+				}
+			}
+		}()
+		wg.Wait()
```

Comment:
Yes that makes sense, I removed the work group in favor of the error check.

Reply:
… actually, do we even need the pull to be in a separate goroutine at all? It could just be on the main path, with a fire-and-forget goroutine created to consume the progress data.

AFAICS the main path currently just blocks on the pull goroutine; it might just as well do the pull instead of waiting.

Reply:
Hah, great catch, I removed that go routine and put it after the metrics collection part.